File: services/crypto-core/state.go
package cryptocore

import (
	"encoding/base64"
	"errors"
	"fmt"
)

type DeviceState struct {
	SigningPrivate  string                        `json:"signingPrivate"`
	SigningPublic   string                        `json:"signingPublic"`
	DHPrivate       string                        `json:"dhPrivate"`
	DHPublic        string                        `json:"dhPublic"`
	SignedPrekey    X25519KeyPairState            `json:"signedPrekey"`
	SignedPrekeySig string                        `json:"signedPrekeySig"`
	OneTime         map[uint32]X25519KeyPairState `json:"oneTime,omitempty"`
	NextOTKID       uint32                        `json:"nextOtkId"`
}

type X25519KeyPairState struct {
	Private string `json:"private"`
	Public  string `json:"public"`
}

type SessionStateSnapshot struct {
	RootKey         string             `json:"rootKey"`
	SendChain       ChainStateSnapshot `json:"sendChain"`
	RecvChain       ChainStateSnapshot `json:"recvChain"`
	RatchetPrivate  string             `json:"ratchetPrivate"`
	RatchetPublic   string             `json:"ratchetPublic"`
	RemoteRatchet   string             `json:"remoteRatchet"`
	RemoteIdentity  string             `json:"remoteIdentity"`
	RemoteSignature string             `json:"remoteSignature"`
	PN              uint32             `json:"pn"`
	Role            SessionRole        `json:"role"`
	PendingPrekey   *uint32            `json:"pendingPrekey,omitempty"`
	Skipped         map[string]string  `json:"skipped,omitempty"`
}

type ChainStateSnapshot struct {
	Key   string `json:"key"`
	Index uint32 `json:"index"`
}

func (d *Device) Export() (*DeviceState, error) {
	if d == nil {
		return nil, errors.New("cryptocore: nil device")
	}
	state := &DeviceState{
		SigningPrivate: base64.StdEncoding.EncodeToString(d.identity.signingPrivate),
		SigningPublic:  base64.StdEncoding.EncodeToString(d.identity.signingPublic),
		DHPrivate:      base64.StdEncoding.EncodeToString(d.identity.dhPrivate[:]),
		DHPublic:       base64.StdEncoding.EncodeToString(d.identity.dhPublic[:]),
		SignedPrekey: X25519KeyPairState{
			Private: base64.StdEncoding.EncodeToString(d.signedPrekey.Private[:]),
			Public:  base64.StdEncoding.EncodeToString(d.signedPrekey.Public[:]),
		},
		SignedPrekeySig: base64.StdEncoding.EncodeToString(d.signedSig),
		OneTime:         make(map[uint32]X25519KeyPairState, len(d.oneTime)),
		NextOTKID:       d.nextOTKID,
	}
	for id, entry := range d.oneTime {
		state.OneTime[id] = X25519KeyPairState{
			Private: base64.StdEncoding.EncodeToString(entry.key.Private[:]),
			Public:  base64.StdEncoding.EncodeToString(entry.key.Public[:]),
		}
	}
	if len(state.OneTime) == 0 {
		state.OneTime = nil
	}
	return state, nil
}

func ImportDevice(state *DeviceState) (*Device, error) {
	if state == nil {
		return nil, errors.New("cryptocore: nil device state")
	}
	signingPriv, err := base64.StdEncoding.DecodeString(state.SigningPrivate)
	if err != nil {
		return nil, fmt.Errorf("cryptocore: decode signing private: %w", err)
	}
	signingPub, err := base64.StdEncoding.DecodeString(state.SigningPublic)
	if err != nil {
		return nil, fmt.Errorf("cryptocore: decode signing public: %w", err)
	}
	dhPriv, err := decodeFixed(state.DHPrivate, 32)
	if err != nil {
		return nil, fmt.Errorf("cryptocore: decode dh private: %w", err)
	}
	dhPub, err := decodeFixed(state.DHPublic, 32)
	if err != nil {
		return nil, fmt.Errorf("cryptocore: decode dh public: %w", err)
	}
	signedPriv, err := decodeFixed(state.SignedPrekey.Private, 32)
	if err != nil {
		return nil, fmt.Errorf("cryptocore: decode signed prekey private: %w", err)
	}
	signedPub, err := decodeFixed(state.SignedPrekey.Public, 32)
	if err != nil {
		return nil, fmt.Errorf("cryptocore: decode signed prekey public: %w", err)
	}
	sig, err := base64.StdEncoding.DecodeString(state.SignedPrekeySig)
	if err != nil {
		return nil, fmt.Errorf("cryptocore: decode signed prekey sig: %w", err)
	}
	dev := &Device{
		identity: identityKeyPair{
			signingPublic:  append([]byte(nil), signingPub...),
			signingPrivate: append([]byte(nil), signingPriv...),
			dhPrivate:      [32]byte{},
			dhPublic:       [32]byte{},
		},
		signedPrekey: keyPair{},
		signedSig:    append([]byte(nil), sig...),
		oneTime:      make(map[uint32]oneTimeEntry),
		nextOTKID:    state.NextOTKID,
	}
	copy(dev.identity.dhPrivate[:], dhPriv)
	copy(dev.identity.dhPublic[:], dhPub)
	copy(dev.signedPrekey.Private[:], signedPriv)
	copy(dev.signedPrekey.Public[:], signedPub)
	for id, kp := range state.OneTime {
		priv, err := decodeFixed(kp.Private, 32)
		if err != nil {
			return nil, fmt.Errorf("cryptocore: decode one-time private: %w", err)
		}
		pub, err := decodeFixed(kp.Public, 32)
		if err != nil {
			return nil, fmt.Errorf("cryptocore: decode one-time public: %w", err)
		}
		var entry keyPair
		copy(entry.Private[:], priv)
		copy(entry.Public[:], pub)
		dev.oneTime[id] = oneTimeEntry{key: entry}
	}
	return dev, nil
}

func ExportSession(state *SessionState) (*SessionStateSnapshot, error) {
	if state == nil {
		return nil, errors.New("cryptocore: nil session")
	}
	snap := &SessionStateSnapshot{
		RootKey:         base64.StdEncoding.EncodeToString(state.RootKey[:]),
		SendChain:       exportChain(state.SendChain),
		RecvChain:       exportChain(state.RecvChain),
		RatchetPrivate:  base64.StdEncoding.EncodeToString(state.RatchetPrivate[:]),
		RatchetPublic:   base64.StdEncoding.EncodeToString(state.RatchetPublic[:]),
		RemoteRatchet:   base64.StdEncoding.EncodeToString(state.RemoteRatchet[:]),
		RemoteIdentity:  base64.StdEncoding.EncodeToString(state.RemoteIdentity[:]),
		RemoteSignature: base64.StdEncoding.EncodeToString(state.RemoteSignature),
		PN:              state.PN,
		Role:            state.Role,
		PendingPrekey:   state.PendingPrekey,
		Skipped:         make(map[string]string, len(state.skipped)),
	}
	for k, v := range state.skipped {
		snap.Skipped[k] = base64.StdEncoding.EncodeToString(v[:])
	}
	if len(snap.Skipped) == 0 {
		snap.Skipped = nil
	}
	return snap, nil
}

func ImportSession(snapshot *SessionStateSnapshot) (*SessionState, error) {
	if snapshot == nil {
		return nil, errors.New("cryptocore: nil session snapshot")
	}
	rootBytes, err := decodeFixed(snapshot.RootKey, 32)
	if err != nil {
		return nil, fmt.Errorf("cryptocore: decode root key: %w", err)
	}
	send, err := importChain(snapshot.SendChain)
	if err != nil {
		return nil, err
	}
	recv, err := importChain(snapshot.RecvChain)
	if err != nil {
		return nil, err
	}
	ratchetPriv, err := decodeFixed(snapshot.RatchetPrivate, 32)
	if err != nil {
		return nil, fmt.Errorf("cryptocore: decode ratchet private: %w", err)
	}
	ratchetPub, err := decodeFixed(snapshot.RatchetPublic, 32)
	if err != nil {
		return nil, fmt.Errorf("cryptocore: decode ratchet public: %w", err)
	}
	remoteRatchet, err := decodeFixed(snapshot.RemoteRatchet, 32)
	if err != nil {
		return nil, fmt.Errorf("cryptocore: decode remote ratchet: %w", err)
	}
	remoteIdentity, err := decodeFixed(snapshot.RemoteIdentity, 32)
	if err != nil {
		return nil, fmt.Errorf("cryptocore: decode remote identity: %w", err)
	}
	remoteSig, err := base64.StdEncoding.DecodeString(snapshot.RemoteSignature)
	if err != nil {
		return nil, fmt.Errorf("cryptocore: decode remote signature: %w", err)
	}
	var root [32]byte
	copy(root[:], rootBytes)
	sess := &SessionState{
		RootKey:   root,
		SendChain: send,
		RecvChain: recv,
	}
	copy(sess.RatchetPrivate[:], ratchetPriv)
	copy(sess.RatchetPublic[:], ratchetPub)
	copy(sess.RemoteRatchet[:], remoteRatchet)
	copy(sess.RemoteIdentity[:], remoteIdentity)
	sess.RemoteSignature = append([]byte(nil), remoteSig...)
	sess.PN = snapshot.PN
	sess.Role = snapshot.Role
	sess.PendingPrekey = snapshot.PendingPrekey
	sess.skipped = make(map[string][32]byte, len(snapshot.Skipped))
	for k, v := range snapshot.Skipped {
		keyBytes, err := decodeFixed(v, 32)
		if err != nil {
			return nil, fmt.Errorf("cryptocore: decode skipped key: %w", err)
		}
		var key [32]byte
		copy(key[:], keyBytes)
		sess.skipped[k] = key
	}
	return sess, nil
}

func exportChain(cs chainState) ChainStateSnapshot {
	return ChainStateSnapshot{
		Key:   base64.StdEncoding.EncodeToString(cs.Key[:]),
		Index: cs.Index,
	}
}

func importChain(cs ChainStateSnapshot) (chainState, error) {
	keyBytes, err := decodeFixed(cs.Key, 32)
	if err != nil {
		return chainState{}, fmt.Errorf("cryptocore: decode chain key: %w", err)
	}
	var key [32]byte
	copy(key[:], keyBytes)
	return chainState{Key: key, Index: cs.Index}, nil
}

func decodeFixed(in string, size int) ([]byte, error) {
	data, err := base64.StdEncoding.DecodeString(in)
	if err != nil {
		return nil, err
	}
	if len(data) != size {
		return nil, fmt.Errorf("unexpected length %d, want %d", len(data), size)
	}
	out := make([]byte, size)
	copy(out, data)
	return out, nil
}

File: frontend/src/crypto-core/state.ts
import { Device } from "./core";
import type {
  ChainState,
  ChainStateSnapshot,
  DeviceState,
  SessionState,
  SessionStateSnapshot,
  SessionRole,
} from "./types";
import { copyBytes, ensureLength, fromBase64, toBase64 } from "./utils";

export function ExportDevice(device: Device): DeviceState {
  if (!device) {
    throw new Error("cryptocore: nil device");
  }
  const state: DeviceState = {
    SigningPrivate: toBase64(device.identity.signingPrivate),
    SigningPublic: toBase64(device.identity.signingPublic),
    DHPrivate: toBase64(device.identity.dhPrivate),
    DHPublic: toBase64(device.identity.dhPublic),
    SignedPrekey: {
      Private: toBase64(device.signedPrekey.Private),
      Public: toBase64(device.signedPrekey.Public),
    },
    SignedPrekeySig: toBase64(device.signedSig),
    OneTime: {},
  };
  for (const [id, entry] of device.oneTime.entries()) {
    state.OneTime![id] = {
      Private: toBase64(entry.key.Private),
      Public: toBase64(entry.key.Public),
    };
  }
  if (state.OneTime && Object.keys(state.OneTime).length === 0) {
    state.OneTime = undefined;
  }
  return state;
}

export function ImportDevice(state: DeviceState): Device {
  if (!state) {
    throw new Error("cryptocore: nil device state");
  }
  const signingPriv = fromBase64(state.SigningPrivate);
  const signingPub = fromBase64(state.SigningPublic);
  const dhPriv = decodeFixed(state.DHPrivate, 32, "dh private");
  const dhPub = decodeFixed(state.DHPublic, 32, "dh public");
  const signedPriv = decodeFixed(state.SignedPrekey.Private, 32, "signed prekey private");
  const signedPub = decodeFixed(state.SignedPrekey.Public, 32, "signed prekey public");
  const sig = fromBase64(state.SignedPrekeySig);
  const identity = {
    signingPublic: signingPub,
    signingPrivate: signingPriv,
    dhPrivate: dhPriv,
    dhPublic: dhPub,
  };
  const device = new Device(identity);
  device.signedPrekey = { Private: signedPriv, Public: signedPub };
  device.signedSig = copyBytes(sig);
  device.oneTime = new Map();
  if (state.OneTime) {
    for (const [key, value] of Object.entries(state.OneTime)) {
      const priv = decodeFixed(value.Private, 32, "one-time private");
      const pub = decodeFixed(value.Public, 32, "one-time public");
      device.oneTime.set(key, { key: { Private: priv, Public: pub } });
    }
  }
  return device;
}

export function ExportSession(state: SessionState): SessionStateSnapshot {
  if (!state) {
    throw new Error("cryptocore: nil session");
  }
  const snapshot: SessionStateSnapshot = {
    RootKey: toBase64(state.RootKey),
    SendChain: exportChain(state.SendChain),
    RecvChain: exportChain(state.RecvChain),
    RatchetPrivate: toBase64(state.RatchetPrivate),
    RatchetPublic: toBase64(state.RatchetPublic),
    RemoteRatchet: toBase64(state.RemoteRatchet),
    RemoteIdentity: toBase64(state.RemoteIdentity),
    RemoteSignature: toBase64(state.RemoteSignature),
    PN: state.PN,
    Role: state.Role as SessionRole,
    PendingPrekey: state.PendingPrekey,
    Skipped: {},
  };
  for (const [k, v] of state.skipped.entries()) {
    snapshot.Skipped![k] = toBase64(v);
  }
  if (snapshot.Skipped && Object.keys(snapshot.Skipped).length === 0) {
    snapshot.Skipped = undefined;
  }
  return snapshot;
}

export function ImportSession(snapshot: SessionStateSnapshot): SessionState {
  if (!snapshot) {
    throw new Error("cryptocore: nil session snapshot");
  }
  const root = decodeFixed(snapshot.RootKey, 32, "root key");
  const send = importChain(snapshot.SendChain);
  const recv = importChain(snapshot.RecvChain);
  const ratchetPriv = decodeFixed(snapshot.RatchetPrivate, 32, "ratchet private");
  const ratchetPub = decodeFixed(snapshot.RatchetPublic, 32, "ratchet public");
  const remoteRatchet = decodeFixed(snapshot.RemoteRatchet, 32, "remote ratchet");
  const remoteIdentity = decodeFixed(snapshot.RemoteIdentity, 32, "remote identity");
  const remoteSig = fromBase64(snapshot.RemoteSignature);
  const session: SessionState = {
    RootKey: root,
    SendChain: send,
    RecvChain: recv,
    RatchetPrivate: ratchetPriv,
    RatchetPublic: ratchetPub,
    RemoteRatchet: remoteRatchet,
    RemoteIdentity: remoteIdentity,
    RemoteSignature: remoteSig,
    PN: snapshot.PN,
    Role: snapshot.Role,
    PendingPrekey: snapshot.PendingPrekey,
    skipped: new Map(),
  };
  if (snapshot.Skipped) {
    for (const [k, v] of Object.entries(snapshot.Skipped)) {
      const keyBytes = decodeFixed(v, 32, "skipped key");
      session.skipped.set(k, keyBytes);
    }
  }
  return session;
}

function exportChain(cs: ChainState): ChainStateSnapshot {
  return {
    Key: toBase64(cs.Key),
    Index: cs.Index,
  };
}

function importChain(cs: ChainStateSnapshot): ChainState {
  return {
    Key: decodeFixed(cs.Key, 32, "chain key"),
    Index: cs.Index,
  };
}

function decodeFixed(input: string, size: number, name: string): Uint8Array {
  const data = fromBase64(input);
  if (data.length !== size) {
    throw new Error(`cryptocore: unexpected length for ${name}: ${data.length}, want ${size}`);
  }
  return ensureLength(data, size, name);
}
